Add account setting to link GitHub account

cl8n opened the pull request:
so open source contributors can link themselves on changelogs

nanaya commented on app/Http/Controllers/Account/GithubUsersController.php:
```diff
@@ -0,0 +1,88 @@
+            'Invalid state.',
+        );
+
+        $tokenResponseBody = (new HttpClient())
```
just use oauth2-github library

nanaya commented on app/Http/Controllers/Account/GithubUsersController.php:
```diff
@@ -0,0 +1,88 @@
+
+        GithubUser::importFromGithub($githubApiUser, auth()->user());
+
+        return redirect(route('account.edit').'#github');
```
I'm not sure about allowing random people to link their github account as opposed to just the ones with actual changelog entry (existing GithubUser from commit callback)

nanaya commented on app/Models/GithubUser.php:
```diff
@@ -3,75 +3,120 @@
+    public function scopeWithGithubInfo(Builder $query): void
     {
-        return $this->hasMany(ChangelogEntry::class);
+        $query->whereNotNull(['canonical_id', 'username']);
```
looking at actual data, there's no entry with either set to null

cl8n replied:
are there any with a blank username? there's some function that looks like it was trying to handle blank or null username

nanaya replied:
nope

nanaya commented on app/Http/Controllers/AccountController.php:
```diff
@@ -113,10 +114,20 @@ public function edit()
 
+        if (GithubUser::canAuthenticate()) {
+            $githubUsers = json_collection(
+                $user->githubUsers()->withGithubInfo()->get(),
```
I'm not sure why I thought multiple github users could be good idea but there's no such thing and I don't think it needs to be supported. It's especially weird when there's actual user interface to link the account(s?).

cl8n replied:
there are at least a few people that have used multiple GitHub accounts on osu-wiki for whatever reason. I agree it's not important but it was easy to support...

nanaya replied:
how many are there? And what's the reason? That just seems confusing.

cl8n replied:
very few. I don't know either lol. anyway looking at it again, yeah it does seem odd to be able to set multiple, I'll make the account page only support linking one

nanaya commented on .env.example:
```diff
@@ -104,6 +104,10 @@ PUSHER_SECRET=
 
 # GITHUB_TOKEN=
 
+# GitHub client for users to associate their GitHub accounts
```
as I learned recently, this should probably also mention where to point the callback url (either here or in setup documentation)

cl8n commented:
oh right I forgot I broke legacy changelog too. i will fix that tomorrow :zzz:

cl8n commented:
... I had a long day :sweat_smile: 

I reverted some of the changes I made to the `GithubUser` model so legacy changelogs can work again (by making fake `GithubUser` with most attributes missing), and added some notes about the potentially-confusing typing

peppy commented:
We were discussing this IRL and not sure how much this makes sense. It's going to end up with a lot of non-developers linking accounts just because the option is there... and the number of contributors over the years is so low we can probably just handle this manually?

It feels like a bit of a waste since this is already implemented though, so curious to hear your thoughts @cl8n.

cl8n commented:
it only lets you link the github account if it has any changelog entries associated. but I could make that more clear in the display as well (https://github.com/ppy/osu-web/pull/9465#discussion_r1241772028)

you probably *could* handle it manually but I don't see why you'd want to... from user perspective I think people generally don't want to bother admins with requests like this either; there's nothing stopping people from asking you to link their account right now but nobody seems to have done that.

(and I will apply last review soon I have a lot of osu catch-up to get through ;_ ; )

peppy commented:
I guess it's okay if it's only for people with contributions (although I don't know how that's checked.. is it post auth?).

It should be shown under the dev section of the account settings so most users don't see it.

cl8n commented on app/Http/Controllers/Account/GithubUsersController.php:
```diff
@@ -0,0 +1,97 @@
+
+        abort_if($user === null, 422, osu_trans('accounts.github_user.error.no_contribution'));
+
+        abort_if($user->user_id !== null, 422, osu_trans('accounts.github_user.error.already_linked'));
```
that check happens in `create()`

peppy commented:
@cl8n is this ready for re-review?

cl8n commented on app/Http/Controllers/Account/GithubUsersController.php:
```diff
@@ -0,0 +1,125 @@
+        abort_if(
+            auth()->user()->githubUser()->exists(),
+            422,
+            'Cannot link more than one GitHub account.',
```
for this one, I changed it so that you get redirected back to the account page (where you should see your existing github account if you didn't already for whatever reason)

for the other untranslated error messages, I left them like that because they should never be seen if not purposely messing with the request from github... seems kinda pointless to translate?
